linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1205,27 +1204,32 @@ private[optimizer] abstract class OptimizerCore(
 
       case PreTransLocalDef(LocalDef(_, _,
           InlineClassInstanceReplacement(_, fieldLocalDefs, cancelFun))) =>
-        val fieldLocalDef = fieldLocalDefs(FieldID(className, field))
+        val fieldLocalDef = fieldLocalDefs(field.name)
         assert(!isLhsOfAssign || fieldLocalDef.mutable, s"assign to immutable field at $pos")
         cont(fieldLocalDef.toPreTransform)
 
       // Select the lo or hi "field" of a Long literal
       case PreTransLit(LongLiteral(value)) if useRuntimeLong =>
         val itemName = field.name
-        assert(itemName == inlinedRTLongLoField ||
-            itemName == inlinedRTLongHiField)
+        assert(itemName.simpleName == inlinedRTLongLoField ||
+            itemName.simpleName == inlinedRTLongHiField)
         assert(expectedType == IntType)
         val resultValue =
-          if (itemName == inlinedRTLongLoField) value.toInt
+          if (itemName.simpleName == inlinedRTLongLoField) value.toInt
           else (value >>> 32).toInt
         cont(PreTransLit(IntLiteral(resultValue)))
 
       case _ =>
         resolveLocalDef(preTransQual) match {
           case PreTransRecordTree(newQual, origType, cancelFun) =>
             val recordType = newQual.tpe.asInstanceOf[RecordType]
-            val recordField = recordType.findField(field.name)
-            val sel = RecordSelect(newQual, field)(recordField.tpe)
+            /* FIXME How come this lookup requires only the `simpleName`?
+             * The `recordType` is created at `InlineableClassStructure.recordType`,
+             * where it uses an allocator. Something fishy is going on here.
+             * (And no, this is not dead code.)
+             */
+            val recordField = recordType.findField(field.name.simpleName)
+            val sel = RecordSelect(newQual, SimpleFieldIdent(recordField.name))(recordField.tpe)
```

Comment:
This snippet is worth paying particular attention to. The change does not alter the logic AFAICT, but it does highlight that there is/was something fishy going on.

Reply:
Looks like records only use simple field names. Maybe this will blow up:

```scala
class A {
  private var x: Int
  def foo = x
}

@inline
class B extends A {
  private var x: String
}

new B().foo
```

Reply:
I have checked this, and it seems to work fine. What I've tested:


```scala
class A {
  private var x: Int = 1
  def foo = x
}

@inline
class B extends A {
  private var x: String = "f"
  def bar = x
}

object HelloWorld {
  @noinline
  def p(x: Any) = println(x)

  def main(args: Array[String]): Unit = {
    val b = new B()

    p(b.foo)
    p(b.bar)
  }
}
```

Relevant generated code:

```js
$c_Lhelloworld_HelloWorld$.prototype.main__AT__V = (function(args) {
  var x = 0;
  var x$1 = null;
  x = 1;
  x$1 = "f";
  this.p__O__V(x);
  this.p__O__V(x$1);
});
```

Reply:
Weird. Perhaps that specific example does not run into that particular code path.